pkg/distributor/distributor.go

```diff
@@ -605,6 +609,79 @@ func (d *Distributor) validateSeries(nowt time.Time, ts mimirpb.PreallocTimeseri
 	return nil
 }
 
+func (d *Distributor) pushWithMiddlewares() push.Func {
+	var middlewares []func(push.Func) push.Func
+
+	// The middlewares will be applied in the order of the slice "middlewares",
+	// requests will traverse them in the reverse order.
+	middlewares = append(middlewares, d.PrePushForwardingMiddleware)
+	middlewares = append(middlewares, d.PrePushHaDedupeMiddleware)
+
+	push := d.PushWithCleanup
+	for _, middleware := range middlewares {
+		push = middleware(push)
+	}
+
+	return push
+}
+
+func (d *Distributor) PrePushHaDedupeMiddleware(next push.Func) push.Func {
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if d.limits.AcceptHASamples(userID) && len(req.Timeseries) > 0 {
+			haReplicaLabel := d.limits.HAReplicaLabel(userID)
+			cluster, replica := findHALabels(haReplicaLabel, d.limits.HAClusterLabel(userID), req.Timeseries[0].Labels)
+			// Make a copy of these, since they may be retained as labels on our metrics, e.g. dedupedSamples.
+			cluster, replica = copyString(cluster), copyString(replica)
+
+			span := opentracing.SpanFromContext(ctx)
+			if span != nil {
+				span.SetTag("cluster", cluster)
+				span.SetTag("replica", replica)
+			}
+
+			numSamples := 0
+			for _, ts := range req.Timeseries {
+				numSamples += len(ts.Samples)
+			}
+			removeReplica := false
+			removeReplica, err = d.checkSample(ctx, userID, cluster, replica)
+			if err != nil {
+				if errors.Is(err, replicasNotMatchError{}) {
+					// These samples have been deduped.
+					d.dedupedSamples.WithLabelValues(userID, cluster).Add(float64(numSamples))
+					return nil, httpgrpc.Errorf(http.StatusAccepted, err.Error())
```

Comment:
If we don't call `next`, then we need to call `cleanup` otherwise it will never be called.

Reply:
nice catch, thx!
fixed d24432ed4597e6558700ce772db7d994fedff442

Reply:
I didn't resolve this comment so you can take a look and review my commit, because this is quite critical and it's easy to miss some detail.

Reply:
I think we should add a unit test for this middleware that also verifies that cleanup is called. WDYT?

Reply:
I added a test which counts how many times cleanup gets called to verify that it gets called for every request: https://github.com/grafana/mimir/pull/2603/commits/6f2a7a1788b1e7000ef10c3ca892fb6121001a07